theseus/optimizer/marginals.py

```diff
@@ -3,23 +3,98 @@
 # This source code is licensed under the MIT license found in the
 # LICENSE file in the root directory of this source tree.
 
-from typing import Sequence
+from itertools import count
+from typing import Optional, Sequence
 
 import torch
 
 from theseus.geometry import Manifold
 
 
-class Marginals:
-    def __init__(self, variables: Sequence[Manifold]):
-        tot_dof = 0
-        for v in variables:
-            tot_dof += v.dof()
-        self.tot_dof = tot_dof
+class Marginal:
+    _ids = count(0)
 
-        self.precision = torch.zeros(
-            self.tot_dof, self.tot_dof, dtype=variables[0].dtype
+    def __init__(
+        self,
+        mean: Sequence[Manifold],
+        precision: Optional[torch.Tensor] = None,
+        name: Optional[str] = None,
+    ):
+        self._id = next(Marginal._ids)
+        if name:
+            self.name = name
+        else:
+            self.name = f"{self.__class__.__name__}__{self._id}"
+
+        dof = 0
+        for v in mean:
+            dof += v.dof()
+        self._dof = dof
+
+        self.mean = mean
+        self.precision = (
+            torch.zeros(mean[0].shape[0], self.dof, self.dof)
+            .to(mean[0].dtype)
+            .to(mean[0].device)
         )
 
+    @property
     def dof(self) -> int:
-        return self.tot_dof
+        return self._dof
+
+    @property
+    def device(self) -> torch.device:
+        return self.precision[0].device
```

Comment:
Should device, dtype, and to also operate over `mean` or just precision? cc @luisenp

Reply:
It seems that `to` operates over mean and precision, which seems OK to me. `device` and `dtype` returning precision's properties seems fine, but maybe we can add a consistency check between `mean` and `precision` in the constructor.